### PR TITLE
[Fix] Ascend copy: fold leading dims of a >2D GM<->UB region into the DMA row count (#1231)

### LINKED ISSUE #1231: [Bug]  T.copy 将3D BufferRegion(16,4,8) flatten为2D (4,8)（取最后2维度），而非总平面 (64,8)。批量DMA只拷了32元素（token 0），其余15个token UB未初始化
## Bug Description
 copy_gm_to_ub 将3D BufferRegion(16,4,8) flatten为2D (4,8)（取最后2维度），而非总平面 (64,8)。批量DMA只拷了32元素（token 0），其余15个token UB未初始化

Image

## Environment

**Hardware:**
- NPU Device: [ A3]

**Software:**
- OS: [EulerOS]
- CANN version: [8.5.0]
- torch-npu version: [2.10.0RC2]
- Python version: [3.11]
- TileLang-Ascend version: [5a221177f184bc52e97620d06007b867af7bc95d]
- Installation method: [build from source]

## Code to Reproduce

```python
# Minimal reproducible code
import tilelang
import torch

FWD_PASS_CONFIGS = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: False,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}

_BWD_PASS_CONFIGS = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_CV_COMBINE: True,
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: True,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: True,
}

VEC_NUM = 2


@tilelang.jit(pass_configs=_FWD_PASS_CONFIGS)
def _mhc_sinkhorn_fwd(
    hidden_size: int,
    token_block_size: int,
    repeat: int,
    eps: float,
) -> tilelang.JITKernel:
    assert hidden_size == 4
    num_tokens = T.symbolic('num_tokens')
    dtype = "float32"
    pad_h = ((hidden_size + 7) // 8) * 8
    sub_block_tokens = token_block_size // VEC_NUM

    @T.prim_func
    def mhc_sinkhorn_kernel(
        comb_res_mix: T.Tensor[(num_tokens, hidden_size, pad_h), dtype],
        comb_res_mix_out: T.Tensor[(num_tokens, hidden_size, hidden_size), dtype],
    ) -> None:
        with T.Kernel(T.ceildiv(num_tokens, token_block_size), is_npu=True) as (cid, vid):
            comb_ub = T.alloc_ub((sub_block_tokens, hidden_size, pad_h), dtype)
            row_max_col = T.alloc_ub((hidden_size, 1), dtype)
            row_max_2d = T.alloc_ub((hidden_size, pad_h), dtype)
            row_sum_col = T.alloc_ub((hidden_size, 1), dtype)
            row_sum_2d = T.alloc_ub((hidden_size, pad_h), dtype)
            col_sum_ub = T.alloc_ub((pad_h,), dtype)
            col_sum_2d = T.alloc_ub((hidden_size, pad_h), dtype)

            row_start = cid * token_block_size + vid * sub_block_tokens

            with T.Scope("V"):
                T.set_flag("mte3", "mte2", 0)
                T.wait_flag("mte3", "mte2", 0)

                for i_tok in T.serial(sub_block_tokens):
                    token_idx = row_start + i_tok
                    if token_idx < num_tokens:
                        T.copy(
                            comb_res_mix[token_idx, :, :],
                            comb_ub[i_tok, :, :],
                        )

                T.set_flag("mte2", "v", 0)
                T.wait_flag("mte2", "v", 0)

                for i_tok in T.serial(sub_block_tokens):
                    token_idx = row_start + i_tok
                    if token_idx < num_tokens:
                        T.reduce_max(comb_ub[i_tok, :, :], row_max_col, dim=-1)
                        T.tile.broadcast(row_max_2d, row_max_col)
                        T.tile.sub(comb_ub[i_tok, :, :], comb_ub[i_tok, :, :], row_max_2d)
                        T.tile.exp(comb_ub[i_tok, :, :], comb_ub[i_tok, :, :])

                        T.reduce_sum(comb_ub[i_tok, :, :], row_sum_col, dim=-1)
                        T.tile.broadcast(row_sum_2d, row_sum_col)
                        T.tile.add(row_sum_2d, row_sum_2d, eps)
                        T.tile.div(comb_ub[i_tok, :, :], comb_ub[i_tok, :, :], row_sum_2d)

                        T.reduce_sum(comb_ub[i_tok, :, :], col_sum_ub, dim=0)
                        T.tile.add(col_sum_ub, col_sum_ub, eps)
                        T.tile.broadcast(col_sum_2d, col_sum_ub)
                        T.tile.div(comb_ub[i_tok, :, :], comb_ub[i_tok, :, :], col_sum_2d)

                        for _ in T.serial(repeat - 1):
                            T.reduce_sum(comb_ub[i_tok, :, :], row_sum_col, dim=-1)
                            T.tile.broadcast(row_sum_2d, row_sum_col)
                            T.tile.add(row_sum_2d, row_sum_2d, eps)
                            T.tile.div(comb_ub[i_tok, :, :], comb_ub[i_tok, :, :], row_sum_2d)

                            T.reduce_sum(comb_ub[i_tok, :, :], col_sum_ub, dim=0)
                            T.tile.add(col_sum_ub, col_sum_ub, eps)
                            T.tile.broadcast(col_sum_2d, col_sum_ub)
                            T.tile.div(comb_ub[i_tok, :, :], comb_ub[i_tok, :, :], col_sum_2d)

                T.set_flag("v", "mte3", 0)
                T.wait_flag("v", "mte3", 0)

                for i_tok in T.serial(sub_block_tokens):
                    token_idx = row_start + i_tok
                    if token_idx < num_tokens:
                        T.copy(comb_ub[i_tok, 0:hidden_size, 0:hidden_size], comb_res_mix_out[token_idx, 0:hidden_size, 0:hidden_size])

                T.set_flag("mte3", "mte2", 0)
                T.wait_flag("mte3", "mte2", 0)

    return mhc_sinkhorn_kernel
```

## Error Message

```
Paste the full error message here
```

## Actual Behavior
A clear and concise description of what actually happened.

## Additional Context

**Programming Mode:**
- [ ] Developer mode (automatic sync/buffer management)
- [ ] Expert mode (manual T.Scope/T.barrier_all)
- [ ] Mixed mode

**Backend:**
- [ ] Ascend C & PTO backend
- [ ] AscendNPU IR backend

**Pass Configs (if applicable):**
```python
pass_configs = {
    tilelang.PassConfigKey.TL_ASCEND_AUTO_SYNC: ...,
    tilelang.PassConfigKey.TL_ASCEND_MEMORY_PLANNING: ...,
    # ...
}
```

**Operator Type:**
[e.g. GEMM, Flash Attention, Softmax, Convolution, etc.]

**Additional Information:**
- Have you checked the [Programming Guide](../docs/TileLang-Ascend%20Programming%20Guide.md)?
- Have you searched existing issues?
- Any other context or screenshots?

### PR DESCRIPTION
**Pull request by 2Sheldonlaw**
Fixes #1231

## Root cause

`copy_gm_to_ub` / `copy_ub_to_gm` 的 DMA 搬运模型是**二维矩形**：`maskShapeM` 个 burst、
每个 `maskShapeN` 个元素，GM 侧行距 `strideN`、UB 侧行距为模板参数（buffer 最后一维）。
`AscendCopy::Lower`（`src/op/ascend.cc`）用 `find_active_dim_indices` 从 region 的活跃维
（extent≠1 的维）里**只取最后两个**当行/列：

- `validRow = ext[倒数第二个活跃维]`
- `validCol = ext[最后一个活跃维]`

对超过 2 个活跃维的 region，前导维度被**静默丢弃**：3D 块拷贝
`A[t:t+16, :, :]`（region (16, 4, 8)，共 512 元素）被降级为 (4, 8)——DMA 只拷 32 个元素
（恰好 token 0），UB tile 其余 15 个 token 未初始化，后续计算读到垃圾值。GM→UB 与
UB→GM 两个方向同病。issue 截图的观察完全准确："将 3D BufferRegion(16,4,8) flatten 为
2D (4,8)（取最后 2 维），而非总平面 (64,8)"。

最小复现（Ascend910 / CANN 9.0.0）：

```python
a_ub = T.alloc_ub((16, 4, 8), "float32")
T.copy(A[row_base : row_base + 16, :, :], a_ub[:, :, :])   # A: (num_tokens, 4, 8)
T.copy(a_ub[:, :, :], B[row_base : row_base + 16, :, :])
```

修复前 GM→UB 往返 max diff 3.66（每块只有 token 0 的数据正确），生成代码
`copy_gm_to_ub<float, 8, 4>(..., 8, 4, 8, 0.0f)`——4 burst × 8 元素 = 32，与上述分析吻合。

## What this PR does

| 文件 | 改动 |
|---|---|
| `src/op/ascend.cc` | `AscendCopy::Lower` 新增 `fold_leading_row_dims`：region 满足内存均匀性（最外活跃维之外，每个活跃行维的 extent 等于其跨度内 buffer 维度的乘积，即中间维完整覆盖）时，前导维折入行数，得到总平面 `(Π ext[行维], ext[列维])`；tail clamp 只作用于最外维后折叠。同步修正 gm2ub/ub2gm 的模板行参数（`Duplicate` pad 尺寸 / 默认 `maskShapeM`）与尾部 `physRow` 参数（`AscendTailMaskPropagation` 的 tail 模型输入），保持与折叠后 validRow 一致 |
| `testing/python/language/test_tilelang_ascend_language_copy_3d_flatten.py` | 新增 5 个用例：3D 整块 fp32/fp16（issue 原始 shape (16,4,8)）、3D 尾块（num_tokens=100，覆盖运行期 clamp + pad 路径）、4D 多级折叠 (2,8,4,8)，全部 GM→UB→GM 往返数值校验 |

生成代码对比（issue 的 (16,4,8) 块）：

```cpp
// 修复前：只拷 4×8=32 元素（token 0），其余 15 个 token UB 未初始化
tl::ascend::copy_gm_to_ub<float, 8, 4>(a_ub[0], A[(cid * 512)], 8, 4, 8, 0.000000e+00f);

// 修复后：64×8 总平面，尾块运行期 clamp
tl::ascend::copy_gm_to_ub<float, 8, 64>(a_ub[0], A[(cid * 512)], 8,
    (((1 <= ((num_tokens / 16) - cid)) ? 16
       : ((0 < (num_tokens - (cid * 16))) ? (num_tokens - (cid * 16)) : 0)) * 4),
    8, 0.000000e+00f);
```

**边界**：

- ≤2 个活跃维的调用路径**产出字节级不变**（fold 仅在 ≥3 活跃维时介入），既有 2D 调用点零影响；
- 不满足均匀性的 >2D region（中间维部分覆盖，如 `A[t:t+16, 0:2, :]`）单个 2D DMA 本就
  无法表示，维持原行为（拆多笔拷贝超出本 PR 范围）。

## Test results

环境：Ascend910_9362 / CANN 9.0.0 / 本仓 dev 构建，ascendc target。

| 项目 | 结果 |
|---|---|
| 3D 块拷贝最小复现（修复前 → 后） | max diff 3.66 → **0.0** |
| issue 原始 kernel（mhc_sinkhorn，num_tokens=128/100） | PASS，max diff 1.19e-07 |
| 新增回归测试 | 5/5 PASS（约 30s，低于 2 分钟门禁） |
| copy 套件（copy_cv / cast_on_copy / copy_pad_value / copy_runtime_extent） | 41 passed |
| gm_to_ub / alloc_codegen / explicit_tmp / narrow_reduce / sync_insert_vs | 293 passed |
| cvseparate（vid 消减） | 17 passed |
| tail_mask_codegen / memory_planning / tail_block | 348 passed（格式化重建后复跑亦通过） |
| 其余 language 套件全量 | 1064 passed, 1 skipped |
| 示例（reduce_min / broadcast / rms_norm / rms_rope_fused / tail_add） | 全部 Kernel Output Match |

已知无关失败：`test_exception_dump`——本环境 CANN 9.0.0 头文件缺 `acldumpTensorInfo`
等 ACL dump API，编译期即失败，与本 PR 无关。

PTO 说明：两后端消费同一组 `AscendCopy::Lower` 产物（validRow/validCol）；PTO 生成代码
已核验正确（`copy_gm_to_ub_dynamic` 的 Shape/Stride/valid 参数均为折叠后的正确值）。本
开发环境的 PTO 执行对所有 kernel（含 2D 基线）均不可运行（bisheng 未定义
`__DAV_C220_VEC__`，kernel 体被跳过），属预先存在的环境限制，非本 PR 引入。

## 回退方式

revert 单个 commit（`bc7bbcfc`）即可。

### CODE REVIEW

**Review by LLMZhangYC — DISMISSED** [review]
lgtm

File: src/op/ascend.cc [hunk 1]
```diff
@@ -188,6 +188,108 @@ Stmt AscendCopy::Lower(const LowerArgs &T, arith::Analyzer *analyzer) const {
...
+
+  // A region with more than two active dims (e.g. a 3D block copy) must be
+  // flattened into the 2D (rows x cols) DMA rectangle the copy helpers
+  // model. The flatten is only memory-uniform when every active row dim
+  // except the outermost fully covers the buffer dims it spans (including
+  // any singleton dims in between); then the leading dims simply multiply
+  // the row count -- the "total plane" (d0*d1, d2). Without the fold the
+  // leading dims were silently dropped and the DMA copied only the
+  // last-2-dims rectangle, leaving most of the destination uninitialized
+  // (issue #1231).
```
Comment (bingmeiyou) [inline src/op/ascend.cc]: 注释中不需要体现 issue

**Comment by bingmeiyou** [comment]
当前拷贝只支持两个活跃维度,  大于两个维度的拷贝在大部分场景下是无法支持的, 因此整体方案也需要再讨论下

File: src/op/ascend.cc (hunk 1 above)
Comment (2Sheldonlaw) [reply src/op/ascend.cc]: Fixed

**Comment by 2Sheldonlaw** [comment]
> 当前拷贝只支持两个活跃维度, 大于两个维度的拷贝在大部分场景下是无法支持的, 因此整体方案也需要再讨论下

感谢 review。说明一下这个 PR 的实际范围——不涉及 3D 切片，只处理整块搬运：
本 PR 做的事：仅当 region 的中间活跃维完整覆盖对应 buffer 维（即 A[t:t+16, :, :] 这种连续整块）时，把前导维折进行数，按总平面一笔 DMA 搬运（(16,4,8) → (64,8)）。这种形态内存连续，一笔 2D DMA 天然可表达，只是此前 lowering 把前导维丢了导致只搬了 1/16。
没有做的事：中间维部分覆盖的形态（如 A[t:t+16, 1:3, :]）维持原有行为不变——这类 region 一笔 2D DMA 无法表示，需要拆多笔拷贝的机制，不在本 PR 范围内。

File: src/op/ascend.cc [hunk 2]
```diff
@@ -188,6 +188,107 @@ Stmt AscendCopy::Lower(const LowerArgs &T, arith::Analyzer *analyzer) const {
...
+          FoldedRegion &res) {
+        if (active.size() < 3) {
+          return;
+        }
+        // Active row dims: active[0 .. m-2] (outermost first), col:
+        // active[m-1]. A uniform burst pitch requires, for each j in [0, m-3]:
+        //   extents[active[j+1]] == prod(buf->shape[active[j]+1 ..
+        //   active[j+1]])
+        // (the intermediate active dim fully spans its group).
+        for (size_t j = 0; j + 2 < active.size(); ++j) {
```
Comment (ChaoyangJi) [inline src/op/ascend.cc:272]: 折叠条件遗漏最后一行维度与列维度之间的间隔

当前循环只检查activel]到active[j+1]之间的维度，没有检查最后一个行维度与列维度之间的维度。例如shape=T,4,2,8]、region=[16,4,1,8]时active:=[0,1,3],这里会被判定为可折叠并生成64x8的DMA矩形，但第2维实际被固定为一个元素，物理行步长仍是2*8。GM侧的strideN可以表达这个间隔，而UB侧没有对应的stride参数，copy_gm_to_ub/copy_ub_to_gm会把数据连续写入错误位置，导致静默数据错误。请在foldable判断中覆盖最后一行维度到列维度之间的维度，并增加这种4D非连续切片的回归测试。

Comment (2Sheldonlaw) [reply src/op/ascend.cc:272]: 问题已在 49f1fa0a 修复：

**1. 补上跨度检查**：`fold_leading_row_dims` 现在要求行维与列维之间所有维的 shape 乘积为 1，否则拒绝折叠。检查**只对 UB 侧强制**：行距在两侧的表达渠道不对称——GM 侧行距走自由的 `strideN` 运行期参数（您的例子中 strideN=16，GM 侧读取其实是正确的）；UB 侧行距被 `dstN`/`srcN` 模板（最后一维）绑定，表达不了 16，所以只有 UB 侧需要拒绝。

**2. 顺带修复姊妹漏洞**：复查发现折叠此前只由 GM 侧判定，"GM 侧干净、但 UB buffer 形状不同"的场景（如 GM `(T,4,8)` → UB `(16,4,2,8)` 取 `[16,4,1,8]`）仍会折出错误行数。现改为**两侧都 foldable 才折叠**（all-or-nothing），不一致时整体回退未折叠的 2D 形态。

**生成代码对比**（您的反例：GM `(T,4,2,8)` region `[16,4,1,8]` → UB `(16,4,2,8)`）：

```cpp
// 修复前（错误折叠）：64 burst 写错 UB 位置
tl::ascend::copy_gm_to_ub<float, 8, 64>(a_ub[0], A[(cid * 1024)], 16, 64, 8, 0.0f);
// 修复后（拒绝折叠，与基线逐字节一致）
tl::ascend::copy_gm_to_ub<float, 8, 1>(a_ub[0], A[(cid * 1024)], 16, 4, 8, 0.0f);
```

**新增 3 个 4D 回归测试**：

| 用例 | 断言 |
|---|---|
| `test_4d_sliced_copy_must_not_fold` | 您的反例形状，codegen 断言不折叠 |
| `test_4d_sliced_copy_gm_gap_ub_clean_roundtrip` | GM 侧有 16 间隔、UB 侧无（`(16,4,1,8)`）：间隔由 strideN 表达，折叠保持生效且数值往返正确（验证"只查 UB 侧"无误伤） |
| `test_4d_contiguous_full_block_roundtrip` | 连续 4D 整块 `(16,4,2,8)` 折成 `(128,8)` 平面，防矫枉过正 |

回归：本文件 8/8（含原有 5 例），copy 相关套件 151 passed，tail_mask / memory_planning / tail_block 348 passed。
